syntax = "proto2";

package rs;

message CreateSessionRequest {
  optional string address = 1;
}

message DeleteSessionRequest {
}

message SessionHeartBeatRequest {
}

message ExecuteCommandRequest {
  optional string address = 1;
  optional string cmd = 2;
  repeated string args = 3;
  optional int32 wait_for_command = 4 [default = 1];
  optional int32 remote_show_output = 5 [default = 0];
  optional int32 remote_show_error = 6 [default = 0];
  optional int32 local_show_output = 7 [default = 1];
  optional int32 local_show_error = 8 [default = 1];
}

message ExecuteCommandResponse {
  optional string address = 1;
  optional string command = 2;
  optional int32 wait_for_command = 4 [default = 1];
  optional int32 remote_show_output = 5 [default = 0];
  optional int32 remote_show_error = 6 [default = 0];
  optional int32 local_show_output = 7 [default = 1];
  optional int32 local_show_error = 8 [default = 1];
}

message ExecuteCommandOutputRequest {
  optional int64 original_request_id = 1;
  optional int32 is_exit = 2;
  optional int32 exit_code = 3;
  optional int32 is_error_output = 4;
  optional string output = 5;
}

message FileOperationRequest {
  optional string cmd = 1;
  repeated string args = 2;
}

message FileOperationResponse {
  optional string cmd = 1;
  repeated string args = 2;
}

message Request {
  optional string name = 1;
  optional int64 session_id = 2 [default = 0];
  optional int64 request_id = 3 [default = 0]; // auto fill by message sender
  optional int64 srv_uid = 4 [default = 0];

  optional int64 timestamp = 100 [default = 0]; // auto fill by message sender

  optional CreateSessionRequest create_session = 200;
  optional DeleteSessionRequest delete_session = 201;
  optional SessionHeartBeatRequest session_heart_beat = 202;
  
  optional ExecuteCommandRequest execute_command = 300;
  optional ExecuteCommandOutputRequest execute_output = 301;
  optional FileOperationRequest file_operation = 302;
}

message Response {
  optional string name = 1; // auto fill from request
  optional int64 session_id = 2 [default = 0];
  optional int64 error_code = 3;
  optional int64 srv_uid = 4 [default = 0];

  optional int64 timestamp = 100 [default = 0]; // auto fill from request
  
  optional int64 request_timestamp = 200 [default = 0]; // auto fill from request
  optional int64 request_id = 201 [default = 0]; // auto fill from request
  
  optional ExecuteCommandResponse execute_command = 300;
  optional FileOperationResponse file_operation = 302;
}
